nntrainer/src/conv2d_layer.cpp

```diff
@@ -370,6 +493,108 @@ int Conv2DLayer::conv2d(float *in, TensorDim indim, const float *kernel,
     }
     I++;
   }
+  return status;
+}
+
+int Conv2DLayer::conv2d_gemm(const float *mKernel, TensorDim kdim,
+                             Tensor const &in, TensorDim outdim,
+                             unsigned int const *mstride,
+                             unsigned int const *pad, float *out,
+                             unsigned int o_size, bool channel_mode) {
+  int status = ML_ERROR_NONE;
+  std::vector<float> inCol;
+
+  if (channel_mode) {
+    inCol.resize(kdim.getFeatureLen() * outdim.width() * outdim.height());
+  } else {
+    inCol.resize(kdim.width() * kdim.height() * outdim.width());
+  }
+
+  Tensor in_padded = zero_pad(0, in, pad);
+  status = im2col(in_padded, kdim, inCol.data(), outdim, mstride, channel_mode);
+  if (status != ML_ERROR_NONE)
+    throw std::runtime_error("Forwarding Convolution failed.");
+
+  float alpha_dgemm = 1.0f;
+  float beta_dgemm = 0.0f;
+  const float *data = mKernel;
+  const float *mdata = inCol.data();
+  float *rdata = out;
+
+  unsigned int kh, kw, w;
+
+  if (channel_mode) {
+    kh = kdim.batch();
+    kw = kdim.getFeatureLen();
+    w = outdim.width() * outdim.height();
+  } else {
+    kh = outdim.height();
+    kw = kdim.width() * kdim.height();
+    w = outdim.width();
+  }
+
+#ifdef USE_BLAS
+  cblas_sgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, kh, w, kw, alpha_dgemm,
+              data, kw, mdata, w, beta_dgemm, rdata, w);
+#else
+  throw std::runtime_error("Need to build with CBLAS library");
+#endif
+
+  return status;
+}
+
+int Conv2DLayer::im2col(Tensor in_padded, TensorDim kdim, float *inCol,
```

Comment:
Considering that this function is fairly complex, I think we should have a test for this function.

Reply:
another pr I guess. ( Besides, if this is wrong, all the test cases releated with conv2d are failed )